verifier/verifier.go

```diff
@@ -898,3 +903,135 @@ func logVerificationResult(logger log.Logger, result *notation.ValidationResult)
+	}
+	if !tsaEnabled {
+		logger.Info("Timestamp verification disabled: no tsa trust store is configured in trust policy")
+		performTimestampVerification = false
+	}
+
+	// check based on 'verifyTimestamp' field
+	timeOfVerification := time.Now()
+	if performTimestampVerification &&
+		signatureVerification.VerifyTimestamp == trustpolicy.OptionAfterCertExpiry {
+		// check if signing cert chain has expired
+		var expired bool
+		for _, cert := range signerInfo.CertificateChain {
+			if timeOfVerification.After(cert.NotAfter) {
+				expired = true
+				break
+			}
+		}
+		if !expired {
+			logger.Infof("Timestamp verification disabled: verifyTimestamp is set to %q and signing cert chain unexpired", trustpolicy.OptionAfterCertExpiry)
+			performTimestampVerification = false
+		}
+	}
+
+	// timestamp verification disabled, signing cert chain MUST be valid
+	// at time of verification
+	if !performTimestampVerification {
+		for _, cert := range signerInfo.CertificateChain {
+			if timeOfVerification.Before(cert.NotBefore) {
+				return fmt.Errorf("verification time is before certificate %q validity period, it will be valid from %q", cert.Subject, cert.NotBefore.Format(time.RFC1123Z))
+			}
+			if timeOfVerification.After(cert.NotAfter) {
+				return fmt.Errorf("verification time is after certificate %q validity period, it was expired at %q", cert.Subject, cert.NotAfter.Format(time.RFC1123Z))
+			}
+		}
+
+		// success
+		return nil
+	}
+
+	// Performing timestamp verification
+	// 1. Timestamp countersignature MUST be present
+	logger.Info("Checking timestamp countersignature existence...")
+	if len(signerInfo.UnsignedAttributes.TimestampSignature) == 0 {
+		return errors.New("no timestamp countersignature was found in the signature envelope")
+	}
+
+	// 2. Verify the timestamp countersignature
+	logger.Info("Verifying the timestamp countersignature...")
+	signedToken, err := tspclient.ParseSignedToken(signerInfo.UnsignedAttributes.TimestampSignature)
+	if err != nil {
+		return fmt.Errorf("failed to parse timestamp countersignature with error: %w", err)
+	}
+	info, err := signedToken.Info()
+	if err != nil {
+		return fmt.Errorf("failed to get the timestamp TSTInfo with error: %w", err)
+	}
+	trustTSACerts, err := loadX509TSATrustStores(ctx, outcome.EnvelopeContent.SignerInfo.SignedAttributes.SigningScheme, policyName, trustStores, x509TrustStore)
+	if err != nil {
+		return fmt.Errorf("failed to load tsa trust store with error: %w", err)
+	}
+	if len(trustTSACerts) == 0 {
+		return errors.New("no trusted TSA certificate found in trust store")
+	}
+	rootCertPool := x509.NewCertPool()
+	for _, trustedCerts := range trustTSACerts {
+		rootCertPool.AddCert(trustedCerts)
+	}
+	timestamp, err := info.Validate(signerInfo.Signature)
+	if err != nil {
+		return fmt.Errorf("failed to get timestamp from timestamp countersignature with error: %w", err)
+	}
```

Comment:
Should we do this before loading certificate? as its relatively less expensive operation that reading certs

Reply:
Updated.